protocol/v1/qbft/msgqueue/queue.go

```diff
@@ -187,14 +187,20 @@ func (q *queue) Peek(n int, idx Index) []*spectypes.SSVMessage {
 
 // WithIterator looping through all indexes and return true when relevant and pop
 func (q *queue) WithIterator(n int, peek bool, iterator func(index Index) bool) []*spectypes.SSVMessage {
+	q.itemsLock.RLock()
```

Comment:
Using a lock here depends on the iterator, which in our case also uses a lock for caching.

I've created another branch with a different solution of copying the indices into a local slice before working, see branch [msgqueue-concurrent-rw](https://github.com/bloxapp/ssv/blob/msgqueue-concurrent-rw/protocol/v1/qbft/msgqueue/queue.go#L189).

Reply:
@amir-blox I saw that the iterator uses a cache's lock, not a queue's, so I thought that my solution would be more effective if the queue is long. However, now I think your solution is better because it doesn't require an iterator not to use the queue's lock. So let's close my PR and use your solution.

Reply:
I'm running both solutions now, let's wait for some insights

Reply:
Pushed my change, we don't see differences but as mentioned above it would be better to copy indices (just keys, not the entire list of messages))